### PR TITLE
Remove visibility changes in draw for *Cursor widgets

### LINKED ISSUE #19633: Multicursor disappears when not moving on nbagg with useblit=False + burns CPU
### Bug report

**Bug summary**
When on the nbagg backend if you stop moving the mouse the multicursor will disappear. The same example works fine on the qt backend.

Additionally I noticed that when I add the multicursor my cpu usage jumps and the kernel busy indicator constantly flashes on and off. 

Showing the plot without the multicursor:
image
and with the multicursor (just displaying, not interacting with the plot):

image
That usage is pretty stable and my laptop's fan goes wild.

The issue with the dissappearing was originally noticed by @ipcoder in https://github.com/matplotlib/ipympl/issues/306

**Code for reproduction**

```python
%matplotlib nbagg
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.widgets import MultiCursor

t = np.arange(0.0, 2.0, 0.01)
s1 = np.sin(2*np.pi*t)
s2 = np.sin(4*np.pi*t)

fig, (ax1, ax2) = plt.subplots(2, sharex=True)
ax1.plot(t, s1)
ax2.plot(t, s2)

multi = MultiCursor(fig.canvas, (ax1, ax2), color='r', lw=1, useblit=False)
plt.show()
```

**Actual outcome**

Peek 2021-03-03 18-12

and the high CPU usage

**Expected outcome**
Red line doesn't disappear + my CPU doesn't get crushed.

**Matplotlib version**

  * Operating system: Ubuntu
  * Matplotlib version (`import matplotlib; print(matplotlib.__version__)`): '3.3.4.post2456+gfd23bb238'
  * Matplotlib backend (`print(matplotlib.get_backend())`): nbagg
  * Python version: '3.9.1 | packaged by conda-forge | (default, Jan 26 2021, 01:34:10) \n[GCC 9.3.0]'
  * Jupyter version (if applicable): Notebook 6.2.0 - IPython 7.20.0

dev instlal of maptlotlib + conda-forge for the others

### PR DESCRIPTION
**Pull request by QuLogic**
## PR Summary

This can be all handled in the mouse move event handler instead, and prevents triggering extra draws in nbAgg.

Fixes #19633.

Additionally, mark access to said event handlers as deprecated.

## PR Checklist

- [ ] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [n/a] New features are documented, with examples if plot related.
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [n/a] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [x] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

File: lib/matplotlib/widgets.py
Comment (tacaswell) [inline lib/matplotlib/widgets.py:1661]: Why drop this?

Comment (tacaswell) [reply lib/matplotlib/widgets.py:1661]: I think we want to keep this short-circuit to avoid going to `_update` which avoids a `draw_idle` down the line.

Comment (timhoffm) [reply lib/matplotlib/widgets.py:1661]: If that's the case, we might want to clarify that for our future selves with a comment.

Comment (QuLogic) [reply lib/matplotlib/widgets.py:1661]: Well, we want to remove it from here so that the later `.set_visible` is effective. I'll see if we can still skip the `_update`. Alternatively, we should turn the visibility attributes into property setters.

Comment (tacaswell) [reply lib/matplotlib/widgets.py:1661]: Oh, I see now.  We used to _always_ make the lines not visible so if `not self.visible` we could safely bail.

Do we need some more careful logic on this like to call `self._update()` if visible or just made invisible?  I suspect there is now also some path where we can set `self.vertOn` or `self.horizOn` to False and then they never get removed?

Comment (QuLogic) [reply lib/matplotlib/widgets.py:1661]: I re-added a condition on the `_update` calls, which is hopefully sufficient.

**Review by timhoffm — APPROVED** [review]

**Comment by QuLogic** [comment]
Hmm, I tested changing those settings, but it may have been only while the mouse was already over the window.

**Comment by tacaswell** [comment]
Pro: rebased
Con: can still reproduce the issues @ianhi reported with disabling the horizontal line

I'm half inclined to merge this even with that issue as this is still a big step forward.

**Comment by ianhi** [comment]
> I'm half inclined to merge this even with that issue as this is still a big step forward.

:rocket: :rocket: :rocket: :rocket: :rocket: - the power to blit in notebooks awaits!

**Review by tacaswell — APPROVED** [review]

**Comment by tacaswell** [comment]
Given Ian's comment I marked this non-draft and approved.  I'm holding off on merging at @timhoffm 's review is very old now so this should get another set of eyes.

**Comment by QuLogic** [comment]
I followed up on both @ianhi's issues in #24845.

**Pull request merged**
